uat/driver/utils.py

```diff
@@ -0,0 +1,81 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import logging
+import subprocess
+
+import tenacity
+from lightkube import Client
+from lightkube.resources.batch_v1 import Job
+from lightkube.resources.core_v1 import Namespace
+
+log = logging.getLogger(__name__)
+
+
+@tenacity.retry(
+    wait=tenacity.wait_exponential(multiplier=2, min=1, max=10),
+    stop=tenacity.stop_after_attempt(30),
+    reraise=True,
+)
+def assert_namespace_active(
+    client: Client,
+    namespace: str,
+):
+    """Test that the provided namespace is Active.
+
+    Retries multiple times to allow for the K8s namespace to be created and reach Active status.
+    """
+    # raises a 404 ApiError if the namespace doesn't exist
+    ns = client.get(Namespace, namespace)
+    phase = ns.status.phase
+
+    log.info(f"Waiting for namespace {namespace} to become 'Active': phase == {phase}")
+    assert phase == "Active", f"Waited too long for namespace {namespace}!"
+
+
+def _log_before_sleep(retry_state):
+    """Custom callback to log the number of seconds before the next attempt."""
+    next_attempt = retry_state.attempt_number
+    delay = retry_state.next_action.sleep
+    log.info(f"Retrying in {int(delay)} seconds (attempts: {next_attempt})")
+
+
+@tenacity.retry(
+    wait=tenacity.wait_exponential(multiplier=2, min=1, max=64),
+    retry=tenacity.retry_if_not_result(lambda result: result),
+    before_sleep=_log_before_sleep,
+    reraise=True,
```

Comment:
can we add a max retries, or some sort of a max time here?

Reply:
Do we really want that? Do we know how long our tests are gonna be running for? This feels weird given that we will be adding more tests in the `integrations` project that the `driver` should know nothing about. I wouldn't want to have to recalculate and adjust this every time we introduce a new test. Let alone that this is not even user-visible. If we were to go down that road, I'd say we take it one step further and add a "this might take about [3 decades] to complete" kind of message.

Reply:
I think we need something here, otherwise any test that breaks in a "I'll never finish" way will run for 24 hours (github's default timeout for a runner).  They're expensive machines, etc etc, so probably we don't want to waste.

I'd say take the current length of time and double it, then we'll revisit in the future?  

Agreed though that, where this is setting a timeout for the entire job, this is awkward to set.  It would be much nicer if we had per-integration timeouts

Reply:
I can do something like an arbitrary hour for now (with the fact that I'll be adding more tests super soon in mind).